### PR TITLE
[MODEL]add baidu ernie model

### PR DESCRIPTION
**Pull request by paperplanet**
## Description ##
add baidu ernie model

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented

### Changes ###
add baidu ernie model

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-759/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-759/7/index.html

File: src/gluonnlp/model/bert.py [hunk 1]
```diff
@@ -665,6 +702,61 @@ def bert_24_1024_16(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu()
...
+    number of self-attention heads (A) is 12.
+
+    Parameters
+    ----------
+    dataset_name : str or None, default None
+        If not None, the dataset name is used to load a vocabulary for the
+        dataset. If the `pretrained` argument is set to True, the dataset name
+        is further used to select the pretrained parameters to load.
+        The supported datasets are 'ernie'
+        :TODO add dataset
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/bert.py]: remove `:TODO add dataset`

File: src/gluonnlp/model/bert.py [hunk 2]
```diff
@@ -665,6 +702,61 @@ def bert_24_1024_16(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu()
                           pretrained_allow_missing=pretrained_allow_missing, **kwargs)
 
 
+def ernie_12_768_12(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu(),
+                    root=os.path.join(get_home_dir(), 'models'), use_pooler=True, use_decoder=True,
+                    use_classifier=True, pretrained_allow_missing=False, **kwargs):
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/bert.py]: remove pretrained_allow_missing from the API

File: src/gluonnlp/model/bert.py [hunk 3]
```diff
@@ -665,6 +702,61 @@ def bert_24_1024_16(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu()
...
+        raised.
+
+    Returns
+    -------
+    BERTModel, gluonnlp.vocab.BERTVocab
+    """
+    return get_bert_model(model_name='ernie_12_768_12', vocab=vocab, dataset_name=dataset_name,
+                          pretrained=pretrained, ctx=ctx, use_pooler=use_pooler,
+                          use_decoder=use_decoder, use_classifier=use_classifier, root=root,
+                          pretrained_allow_missing=pretrained_allow_missing, **kwargs)
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/bert.py]: pass pretrained_allow_missing=False

File: src/gluonnlp/model/bert.py (hunk 2 above)
Comment (szha) [reply src/gluonnlp/model/bert.py]: Maybe we could just reuse `bert_12_768_12`? the dataset name 'baidu_ernie_uncased' should make it clear that it's the ERNIE model.

**Comment by mli** [comment]
Job PR-759/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-759/8/index.html

File: src/gluonnlp/model/bert.py (hunk 2 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: ERNIE has max_length=513, activation=relu, layer_norm_eps=1e-5 which is a barrier

File: src/gluonnlp/model/bert.py (hunk 1 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: done

**Comment by mli** [comment]
Job PR-759/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-759/9/index.html

**Comment by mli** [comment]
Job PR-759/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-759/10/index.html

**Comment by mli** [comment]
Job PR-759/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-759/13/index.html

File: src/gluonnlp/model/bert.py (hunk 3 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: done

File: src/gluonnlp/model/bert.py [hunk 4]
```diff
@@ -82,6 +83,10 @@ class BERTPositionwiseFFN(BasePositionwiseFFN):
         Prefix for name of `Block`s (and name of weight if params is `None`).
     params : Parameter or None
         Container for weight sharing between cells. Created if `None`.
+    activation : str, default None
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/bert.py]: update doc: default gelu

File: src/gluonnlp/model/bert.py [hunk 5]
```diff
@@ -665,6 +703,50 @@ def bert_24_1024_16(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu()
                           pretrained_allow_missing=pretrained_allow_missing, **kwargs)
 
 
+def ernie_12_768_12(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu(),
+                    root=os.path.join(get_home_dir(), 'models'), use_pooler=True, use_decoder=True,
+                    use_classifier=True, **kwargs):
+    """baidu ERNIE model.
+
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/bert.py]: Could you add a link to their paper: https://arxiv.org/pdf/1904.09223.pdf

File: src/gluonnlp/model/transformer.py [hunk 6]
```diff
@@ -84,6 +87,10 @@ class BasePositionwiseFFN(HybridBlock):
     params : Parameter or None
         Container for weight sharing between cells.
         Created if `None`.
+    activation : str, default 'relu'
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/transformer.py]: activation was already mentioned in line 79

File: src/gluonnlp/model/transformer.py [hunk 7]
```diff
@@ -221,17 +234,26 @@ def __init__(self, attention_cell='multi_head', units=128,
...
         positionwise_ffn = BERTPositionwiseFFN if use_bert else PositionwiseFFN
-        return positionwise_ffn(units=units, hidden_size=hidden_size, dropout=dropout,
-                                use_residual=use_residual, weight_initializer=weight_initializer,
-                                bias_initializer=bias_initializer)
-
+        ffn_params = {'units':units, 'hidden_size':hidden_size, 'dropout':dropout,
+                      'use_residual':use_residual,
+                      'weight_initializer': weight_initializer,
+                      'bias_initializer': bias_initializer}
+        if use_bert:
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/transformer.py]: The fact that layer_norm_eps and activation only take effect if use_bert_ffn=True makes the code harder to maintain. Shall we add these two flags to PositionwiseFFN, too?

File: src/gluonnlp/model/transformer.py [hunk 8]
```diff
@@ -111,7 +118,8 @@ def __init__(self, units=512, hidden_size=2048, dropout=0.0, use_residual=True,
                                   prefix='ffn_2_')
             if dropout:
                 self.dropout_layer = nn.Dropout(rate=dropout)
-            self.layer_norm = _get_layer_norm(use_bert_layer_norm, units)
+            self.layer_norm = _get_layer_norm(use_bert_layer_norm, units,
+                                              layer_norm_eps=layer_norm_eps)
 
     def _get_activation(self, act):
         """Get activation block based on the name. """
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/transformer.py]: what if act is None?

File: src/gluonnlp/model/transformer.py [hunk 9]
```diff
@@ -594,6 +624,10 @@ class TransformerEncoderCell(BaseTransformerEncoderCell):
         Prefix for name of `Block`s. (and name of weight if params is `None`).
     params : Parameter or None
         Container for weight sharing between cells. Created if `None`.
+    activation : str, default None
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/transformer.py]: Should the default value be relu instead? If someone uses TransformerEncoderCell API to contruct a cell directly with the default value, it should set act='relu'. 

Would you mind also adding a unit test for TransformerEncoderCell in https://github.com/dmlc/gluon-nlp/blob/master/tests/unittest/test_models.py#L543 with some dummy inputs, to check if the cell with default arguments can be constructed?

File: src/gluonnlp/model/bert.py [hunk 10]
```diff
@@ -144,6 +150,10 @@ class BERTEncoder(BaseTransformerEncoder):
         Prefix for name of `Block`s. (and name of weight if params is `None`).
     params : Parameter or None
         Container for weight sharing between cells. Created if `None`.
+    activation : str, default None
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/bert.py]: should activation defaults to gelu?

File: src/gluonnlp/model/bert.py [hunk 11]
```diff
@@ -216,6 +228,10 @@ class BERTEncoderCell(BaseTransformerEncoderCell):
...
```
Comment (eric-haibin-lin) [inline src/gluonnlp/model/bert.py]: should activation defaults to gelu?

File: src/gluonnlp/model/transformer.py (hunk 8 above)
Comment (paperplanet) [reply src/gluonnlp/model/transformer.py]: seems None will not be passed to self._get_activation()
`self.activation = self._get_activation(activation) if activation else None `

File: src/gluonnlp/model/transformer.py (hunk 9 above)
Comment (paperplanet) [reply src/gluonnlp/model/transformer.py]: The default behaviour would de decide upon whether he uses BERTPositionwiseFFN or PositionwiseFFN
`positionwise_ffn = BERTPositionwiseFFN if use_bert else PositionwiseFFN`
https://github.com/paperplanet/gluon-nlp/blob/894b60b2e513606a0fa3ff2c55d8756c4eb8f7d7/src/gluonnlp/model/transformer.py#L246

File: scripts/bert/conversion_tools/convert_paddle_to_gluon.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+import paddle.fluid as fluid
+
+from gluonnlp.model import BERTEncoder, BERTModel
+from gluonnlp.model.bert import bert_hparams
+from utils import get_hash, tf_vocab_to_gluon_vocab, load_text_vocab
+
+
+BAIDU_LARK_PATH = '../../../../LARK'
+if not os.path.exists(BAIDU_LARK_PATH):
+    os.system('git clone https://github.com/PaddlePaddle/LARK.git ' + BAIDU_LARK_PATH)
```
Comment (leezu) [inline scripts/bert/conversion_tools/convert_paddle_to_gluon.py]: Please make `BAIDU_LARK_PATH` a command line argument or change it to the current folder. To keep consistency with the other conversion scripts, maybe don't automatically clone the git repo?

File: src/gluonnlp/model/bert.py [hunk 13]
```diff
@@ -534,9 +553,28 @@ def _decode(self, sequence, masked_positions):
     'word_embed': None,
 }
 
+ernie_12_768_12_hparams = {
+    'attention_cell': 'multi_head',
+    'num_layers': 12,
+    'units': 768,
+    'hidden_size': 3072,
+    'max_length': 513,
```
Comment (leezu) [inline src/gluonnlp/model/bert.py]: Are you aware why they increase from 512 to 513?

File: src/gluonnlp/model/bert.py [hunk 14]
```diff
@@ -665,6 +703,50 @@ def bert_24_1024_16(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu()
                           pretrained_allow_missing=pretrained_allow_missing, **kwargs)
 
 
+def ernie_12_768_12(dataset_name=None, vocab=None, pretrained=True, ctx=mx.cpu(),
+                    root=os.path.join(get_home_dir(), 'models'), use_pooler=True, use_decoder=True,
+                    use_classifier=True, **kwargs):
+    """baidu ERNIE model.
```
Comment (leezu) [inline src/gluonnlp/model/bert.py]: Shoud be "Baidu" not "baidu"

File: src/gluonnlp/model/bert.py (hunk 4 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: updated

File: src/gluonnlp/model/bert.py (hunk 5 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: added

File: src/gluonnlp/model/transformer.py (hunk 6 above)
Comment (paperplanet) [reply src/gluonnlp/model/transformer.py]: ok, deleted

File: src/gluonnlp/model/transformer.py (hunk 7 above)
Comment (paperplanet) [reply src/gluonnlp/model/transformer.py]: Yes, added

File: src/gluonnlp/model/transformer.py (hunk 9 above)
Comment (paperplanet) [reply src/gluonnlp/model/transformer.py]: Yes.
Added a simple test.

File: src/gluonnlp/model/bert.py (hunk 10 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: yes

File: src/gluonnlp/model/bert.py (hunk 11 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: yeah, modified

File: scripts/bert/conversion_tools/convert_paddle_to_gluon.py (hunk 12 above)
Comment (paperplanet) [reply scripts/bert/conversion_tools/convert_paddle_to_gluon.py]: OK. It's done.

File: src/gluonnlp/model/bert.py (hunk 13 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: Not aware why they did so.

File: src/gluonnlp/model/bert.py (hunk 14 above)
Comment (paperplanet) [reply src/gluonnlp/model/bert.py]: Yeah, modified

**Comment by mli** [comment]
Job PR-759/14 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-759/14/index.html

**Comment by mli** [comment]
Job PR-759/15 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-759/15/index.html

**Pull request merged**